File: leetcode-java/src/main/java/chapter03/BinaryTreeRightSideView.java
package chapter03;

import utils.TreeNode;

import java.util.*;

/**
 * 二叉树的右/左视图
 */
public class BinaryTreeRightSideView {

    /**
     * 深度优先, 递归实现
     */
    public static List<Integer> rightSideViewDfs(TreeNode root) {
        Map<Integer, Integer> views = new HashMap<>();
        rightSideViewDfs(root, 0, views);
        return new ArrayList<>(views.values());
    }

    public static void rightSideViewDfs(TreeNode root, int level, Map<Integer, Integer> views) {
        if (root == null) {
            return;
        }
        views.putIfAbsent(level, root.value);
        rightSideViewDfs(root.right, level + 1, views);
        rightSideViewDfs(root.left, level + 1, views);
    }

    /**
     * 广度优先, 迭代实现
     */
    public static List<Integer> rightSideViewBfs(TreeNode root) {
        TreeNode view = null;
        TreeNode currentLevelEnd = root;
        TreeNode nextLevelEnd = null;
        List<Integer> views = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        if (root != null) {
            queue.offer(root);
        }
        while (!queue.isEmpty()) {
            root = queue.poll();
            view = view != null ? view : root;
            if (root.left != null) {
                queue.offer(root.left);
                nextLevelEnd = root.left;
            }
            if (root.right != null) {
                queue.offer(root.right);
                nextLevelEnd = root.right;
            }
            if (currentLevelEnd == root) {
                views.add(view.value);
                currentLevelEnd = nextLevelEnd;
                view = null;
            }
        }
        return views;
    }
}

File: leetcode-go/chapter03/binary_tree_right_side_view.go
package chapter03

import (
	"github.com/emirpasic/gods/queues/linkedlistqueue"
	nodes "leetcode-go/node"
)

// 二叉树的右/左视图
type BinaryTreeRightSideView struct {}

// 深度优先, 递归实现
func (tree *BinaryTreeRightSideView) RightSideViewDfs(root *nodes.TreeNode) []int {
	views := make([]int, 0)
	levels := make(map[int]int)
	var recur func(node *nodes.TreeNode, level int, levels map[int]int)
	recur = func(node *nodes.TreeNode, level int, levels map[int]int) {
		if node == nil {
			return
		}
		if _, ok := levels[level]; !ok {
			levels[level] = node.Val
		}
		recur(node.Right, level + 1, levels)
		recur(node.Left, level + 1, levels)
	}
	recur(root, 0, levels)
	for index := 0; index < len(levels); index++ {
		views = append(views, levels[index])
	}
	return views
}

// 层序遍历, 广度优先, 迭代实现
func (tree *BinaryTreeRightSideView) RightSideViewBfs(root *nodes.TreeNode) []int {
	var currentLevelEnd *nodes.TreeNode = root
	var nextLevelEnd, rightSideView *nodes.TreeNode = nil, nil
	rightSideViews := make([]int, 0)
	queue := linkedlistqueue.New()
	if root != nil {
		queue.Enqueue(root)
	}
	for !queue.Empty() {
		element, _ := queue.Dequeue()
		root = element.(*nodes.TreeNode)
		if rightSideView == nil {
			rightSideView = root
		}
		if root.Right != nil {
			queue.Enqueue(root.Right)
			nextLevelEnd = root.Right
		}
		if root.Left != nil {
			queue.Enqueue(root.Left)
			nextLevelEnd = root.Left
		}
		if currentLevelEnd == root {
			rightSideViews = append(rightSideViews, rightSideView.Val)
			rightSideView = nil
			currentLevelEnd = nextLevelEnd
		}
	}
	return rightSideViews
}
